sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -167,18 +172,80 @@ private TransportCollectProfileOperation getRemoteCollectOperation(DependencyCar
     }
 
-    private Row buildResponse(Map<String, Object> apeTimings, Map<String, Map<String, Object>> nodeTimings) {
+    private Row buildResponse(Map<String, Object> apeTimings,
+                              Map<String, Map<String, Object>> timingsByNodeId,
+                              NodeOperationTree operationTree) {
         MapBuilder<String, Object> mapBuilder = MapBuilder.newMapBuilder();
         apeTimings.forEach(mapBuilder::put);
 
+        Map<String, Object> phasesTimings = extractPhasesTimingsFrom(timingsByNodeId, operationTree);
+        Map<String, Map<String, Object>> resultNodeTimings = getNodeTimingsWithoutPhases(phasesTimings.keySet(), timingsByNodeId);
         MapBuilder<String, Object> executionTimingsMap = MapBuilder.newMapBuilder();
-        nodeTimings.forEach(executionTimingsMap::put);
+        executionTimingsMap.put("Phases", phasesTimings);
+        resultNodeTimings.forEach(executionTimingsMap::put);
         executionTimingsMap.put("Total", apeTimings.get(Phase.Execute.name()));
 
         mapBuilder.put(Phase.Execute.name(), executionTimingsMap.immutableMap());
         return new Row1(mapBuilder.immutableMap());
     }
 
+    private static Map<String, Object> extractPhasesTimingsFrom(Map<String, Map<String, Object>> timingsByNodeId,
+                                                                NodeOperationTree operationTree) {
+        Map<String, Object> allPhases = new TreeMap<>();
+        for (NodeOperation operation : operationTree.nodeOperations()) {
+            ExecutionPhase phase = operation.executionPhase();
+            getPhaseTimingsAndAddThemToPhasesMap(phase, timingsByNodeId, allPhases);
+        }
+
+        ExecutionPhase leafExecutionPhase = operationTree.leaf();
+        getPhaseTimingsAndAddThemToPhasesMap(leafExecutionPhase, timingsByNodeId, allPhases);
+
+        return allPhases;
+    }
+
+    private static void getPhaseTimingsAndAddThemToPhasesMap(ExecutionPhase leafExecutionPhase,
+                                                             Map<String, Map<String, Object>> timingsByNodeId,
+                                                             Map<String, Object> allPhases) {
+        String phaseName = ProfilingContext.generateProfilingKey(leafExecutionPhase.phaseId(), leafExecutionPhase.name());
+        Map<String, Object> phaseTimingsAcrossNodes = getPhaseTimingsAcrossNodes(phaseName, timingsByNodeId);
+
+        if (!phaseTimingsAcrossNodes.isEmpty()) {
+            allPhases.put(phaseName, MapBuilder.newMapBuilder().put("nodes", phaseTimingsAcrossNodes).immutableMap());
+        }
+    }
+
+    private static Map<String, Object> getPhaseTimingsAcrossNodes(String phaseName,
+                                                                  Map<String, Map<String, Object>> timingsByNodeId) {
+        MapBuilder<String, Object> timingsForPhaseAcrossNodes = MapBuilder.newMapBuilder();
+        for (String operationNode : timingsByNodeId.keySet()) {
```

Comment:
Why don't you use an `Entry<String, Map<String, Object>>` iterator on `entrySet()` instead?